libsolidity/analysis/TypeChecker.cpp

```diff
@@ -561,13 +584,12 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 			m_errorReporter.fatalTypeError(var->location(), "Internal or recursive type is not allowed for public or external functions.");
 		if (
 			_function.visibility() > FunctionDefinition::Visibility::Internal &&
-			type(*var)->category() == Type::Category::Struct &&
-			!type(*var)->dataStoredIn(DataLocation::Storage) &&
-			!_function.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2)
+			!_function.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2) &&
+			!typeSupportedByOldABIEncoder(*type(*var))
 		)
 			m_errorReporter.typeError(
 				var->location(),
-				"Structs are only supported in the new experimental ABI encoder. "
+				"This type is only supported in the new experimental ABI encoder. "
```

Comment:
Were the current (not the new ones) tests only checking for "only supported" or we didn't had tests to cover this?

Reply:
We only had tests for structs, not structs inside arrays or arrays containing arrays etc.

Reply:
Yeah but none of the test expectations had to change so either there was no text or it didn't expect the entire message.

Reply:
It seems we did not have tests for that.

Reply:
Is this a blocker for merging? Please make clear statements.

Reply:
Grepped the sources, there was no test for this before.

Reply:
No, there are tests now hopefully covering all of it.